src/cunumeric/index/choose_template.inl

```diff
@@ -0,0 +1,67 @@
+/* Copyright 2021 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "cunumeric/pitches.h"
+
+namespace cunumeric {
+
+using namespace Legion;
+using namespace legate;
+
+template <VariantKind KIND, LegateTypeCode CODE, int DIM>
+struct ChooseImplBody;
+
+template <VariantKind KIND>
+struct ChooseImpl {
+  template <LegateTypeCode CODE, int DIM>
+  void operator()(ChooseArgs& args) const
+  {
+    using VAL = legate_type_of<CODE>;
+    auto rect = args.out.shape<DIM>();
+
+    Pitches<DIM - 1> pitches;
+    size_t volume = pitches.flatten(rect);
+    if (volume == 0) return;
+
+    auto out       = args.out.write_accessor<VAL, DIM>(rect);
+    auto index_arr = args.inputs[0].read_accessor<int, DIM>(rect);
+
+#ifndef LEGION_BOUNDS_CHECKS
+    // Check to see if this is dense or not
+    bool dense =
+      index_arr.accessor.is_dense_row_major(rect) && out.accessor.is_dense_row_major(rect);
+#else
+    // No dense execution if we're doing bounds checks
+    bool dense = false;
+#endif
+    std::vector<AccessorRO<VAL, DIM>> choices;
+    auto rect_c = args.inputs[1].shape<DIM>();
+    for (int i = 1; i < args.inputs.size(); i++) {
+      choices.push_back(args.inputs[i].read_accessor<VAL, DIM>(rect_c));
+      dense = dense && choices[i - 1].accessor.is_dense_row_major(rect_c);
```

Comment:
Correction on the above: the dense-ness check should be done on the rectangle that will eventually be used to index over the arrays (`rect` in this case).

Consider this operation on 4 processors:

```
x = ones((6,1)) + zeros((1,4))
```

What we're trying to check is whether it's safe to do this:

```
for flatidx = 0 upto volume(sub-tile-rect):
  out.ptr[flatidx] = rhs1.ptr[flatidx] + rhs2.ptr[flatidx]
```

instead of this:

```
for p in sub-tile-rect:
  out.accessor[p] = rhs1.accessor[p] + rhs2.accessor[p]
```

For this operation it is not safe to do this. Each point task will get a `(3,2)` sub-tile of the output. While the 6 elements of the output sub-tile are laid out sequentially in memory, the rhs1 sub-tile only has 3 elements (the second column is broadcasted, i.e. it doesn't exist), so it would NOT be OK to use the `flatidx` codepath. We will need to use the full accessor codepath, which will encapsulate all the necessary math to make sure that both `(0,0)` and `(0,1)` dereference the same element of rhs1.

To make sure we test for this properly, we should check the denseness of each array against the sub-shape `(3,2)` that we will use in the access loop later.